src/librustc/traits/select.rs

```diff
@@ -1071,13 +1071,18 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             if !candidate_set.ambiguous && candidate_set.vec.is_empty() {
                 let trait_ref = stack.obligation.predicate.skip_binder().trait_ref;
                 let self_ty = trait_ref.self_ty();
-                let cause = IntercrateAmbiguityCause::UpstreamCrateUpdate {
-                    trait_desc: trait_ref.to_string(),
-                    self_desc: if self_ty.has_concrete_skeleton() {
-                        Some(self_ty.to_string())
-                    } else {
-                        None
-                    },
+                let trait_desc = trait_ref.to_string();
+                let self_desc = if self_ty.has_concrete_skeleton() {
+                    Some(self_ty.to_string())
+                } else {
+                    None
+                };
+                let cause = if
+                    trait_ref.def_id.krate != LOCAL_CRATE &&
+                    !self.tcx().has_attr(trait_ref.def_id, "fundamental") {
```

Comment:
Nit: This feels a bit like a DRY violation -- this same logic appears in `traits::trait_ref_is_knowable()` -- maybe factor it out into a helper? Something like `trait_ref_permits_any_impl`?